chirp/drivers/retevis_ra87.py

```diff
@@ -0,0 +1,1250 @@
+            txmode = "DTCS"
+            mem.dtcs = self._get_dcs(_mem.txtone)
+            if _mem.txdtcs_pol == 1:
+                dtcs_pol[0] = "R"
+
+        if txmode == "Tone" and not rxmode:
+            mem.tmode = "Tone"
+        elif txmode == rxmode and txmode == "Tone" and mem.rtone == mem.ctone:
+            mem.tmode = "TSQL"
+        elif txmode == rxmode and txmode == "DTCS" and mem.dtcs == mem.rx_dtcs:
+            mem.tmode = "DTCS"
+        elif rxmode or txmode:
+            mem.tmode = "Cross"
+            mem.cross_mode = "%s->%s" % (txmode, rxmode)
+
+        mem.dtcs_polarity = "".join(dtcs_pol)
+
+        _levels = self.POWER_LEVELS
+        if _mem.txpower == TXPOWER_HIGH:
+            mem.power = _levels[4]
+        elif _mem.txpower == TXPOWER_MID:
+            mem.power = _levels[3]
+        elif _mem.txpower == TXPOWER_LOW3:
+            mem.power = _levels[2]
+        elif _mem.txpower == TXPOWER_LOW2:
+            mem.power = _levels[1]
+        elif _mem.txpower == TXPOWER_LOW:
+            mem.power = _levels[0]
+        else:
+            LOG.error('%s: get_mem: unhandled power level: 0x%02x' %
+                      (mem.name, _mem.txpower))
+
+        mem.extra = RadioSettingGroup("Extra", "extra")
+        rs = RadioSettingValueBoolean(_mem.beatshift)
+        rset = RadioSetting("beatshift", "Beat Shift", rs)
+        mem.extra.append(rset)
+
+        rs = RadioSettingValueBoolean(_mem.compand)
+        rset = RadioSetting("compand", "Compand", rs)
+        mem.extra.append(rset)
+
+        options = ['Off', 'Freq 1', 'Freq 2', 'Freq 3',
+                   'Freq 4', 'Freq 5', 'Freq 6', 'User']
+        rs = RadioSettingValueList(options, options[_mem.scramble])
+        rset = RadioSetting("scramble", "Scramble", rs)
+        mem.extra.append(rset)
+
+        rs = RadioSettingValueBoolean(_mem.hide)
+        rset = RadioSetting("hide", "Hide Channel", rs)
+        mem.extra.append(rset)
+
+        rs = RadioSettingValueBoolean(_mem.reverse)
+        rset = RadioSetting("reverse", "Reverse", rs)
+        mem.extra.append(rset)
+
+        return mem
+
+    def set_memory(self, mem):
+        # Get a low-level memory object mapped to the image
+        _mem = self._memory_obj()[mem.number]
+
+        if mem.empty:
+            _mem.set_raw(b"\xFF" * 31 + b"\x80")
+
+            return
+
+        _mem.set_raw(b"\x00" * 25 + b"\xFF" * 6 + b"\x00")
+
+        _mem.rxfreq = mem.freq
+
+        if mem.duplex == 'split':
+            _mem.txfreq = mem.offset / 10
+        elif mem.duplex == '+':
+            _mem.duplex = DUPLEX_POSSPLIT
+            _mem.offset = mem.offset / 1000
+        elif mem.duplex == '-':
+            _mem.duplex = DUPLEX_NEGSPLIT
+            _mem.offset = mem.offset / 1000
+        elif mem.duplex == '':
+            _mem.duplex = DUPLEX_NOSPLIT
+        else:
+            LOG.error('%s: set_mem: unhandled duplex: %s' %
+                      (mem.name, mem.duplex))
```

Comment:
The UI should not send you any duplexes other than those your radio claims to handle... Is this for a specific behavior you saw, or are you just being comprehensive? If it can happen for some reason, this should probably raise an error so there's some indication to the user that something went wrong, right? Otherwise it'll just flip back when it refreshes.

Reply:
This is basically code borrowed from the anytone778uv.py driver. DUPLEX_NOSPLIT is just a variable that holds the value of 0x00 which is the value of 'duplex' when the radio's memory is set to simplex (no split).

    u8   step:4,          // STEP                                    // 5
         unk1:2,
         duplex:2;        // Duplex 0: Simplex, 1: Plus, 2: Minus

Do you just want me to rename the variable to something like DUPLEX_SIMPLEX, replace it directly with 0x00 or am I not understanding what you are trying to tell me?

Reply:
Nope, I'm just wondering about the else condition for `mem.duplex` and the error log. I'm just wondering if you added that because you saw the UI send you a duplex that wasn't in the radio's supported_duplexes?

Reply:
No. It it there only because it was in the anytone778uv.py driver that I borrowed it from.

Reply:
Okay, gotcha, just wanted to make sure it wasn't signaling something else going on.

Reply:
There is a similar else condition in both get_memory() and set_memory(). Should I just remove them both?

Reply:
No need to remove them, I just wanted to make sure there wasn't some reason for it :)